### ISSUE TITLE
mapAccumRight argument order?

### ISSUE THREAD

**Issue by tacomanator** [ramda/ramda#2386]
Is the argument order for `mapAccumRight` correct? Docs say 

> iterator function receives two arguments, value and acc, and should return a tuple [value, acc]

But using as documented:

```javascript
var iterator = (value, acc) => [value, acc];
R.mapAccumRight(iterator, "acc", ["a", "b"]); //=> [["b", "acc"], "a"]
```

Reversing order of tuple returned by iterator works as expected:
```javascript
var iterator = (value, acc) => [acc, value];
R.mapAccumRight(iterator, "acc", ["a", "b"]); //=> [["a", "b"], "acc"]

**Comment by kedashoe** [comment]
Can't say I understand this function well, but it seems the parameters are being passed along as advertised: https://goo.gl/XgrztS

```js
var iterator = (value, acc) => (console.log(`val=${value}, acc=${acc}`), [value, acc]);
R.mapAccumRight(iterator, "acc", ["a", "b", "c"]);

//=> val=c, acc=acc
//=> val=b, acc=c
//=> val=a, acc=b
```

**Comment by tacomanator** [comment]
@kedashoe I could be misunderstanding but I don't think that it is. Given that your iterator does not change acc, just passes it along, shouldn't the output instead be as follows?
```javascript
//=> val=c, acc=acc
//=> val=b, acc=acc
//=> val=a, acc=acc
```

That's how `mapAccum` works, just in the opposite orderin the opposite order (argument order for iterator and tuple reversed per docs for `mapAccum`):

```javascript
var iterator = (acc, value) => (console.log(`val=${value}, acc=${acc}`),[acc, value]);
R.mapAccum(iterator, "acc", ["a", "b", "c"]);

val=a, acc=acc
val=b, acc=acc
val=c, acc=acc
```

**Comment by kedashoe** [comment]
Looks to me like there is an issue. We brought this over from [Haskell](http://hackage.haskell.org/package/base-4.7.0.1/docs/Data-List.html#v:mapAccumR). We changed what this function does in https://github.com/ramda/ramda/pull/1975. Haskell's version has the same parameter order for the Left and Right versions, and iterates the list is the same order, but applies the accumulator differently.

```haskell
-- haskell
Prelude Data.List> mapAccumR (\acc val -> (acc ++ "x", val ++ acc)) "x" ["a", "b", "c"]
("xxxx",["axxx","bxx","cx"])
Prelude Data.List> mapAccumL (\acc val -> (acc ++ "x", val ++ acc)) "x" ["a", "b", "c"]
("xxxx",["ax","bxx","cxxx"])
```
```js
// js
R.mapAccumRight((val, acc) => [val + acc, acc + "x"], "x", ["a", "b", "c"])
//=> [["bcxx", "cxx", "xx"], "abcx"]
R.mapAccum((acc, val) => [acc + "x", val + acc], "x", ["a", "b", "c"])
//=> ["xxxx", ["ax", "bxx", "cxxx"]]
```
Thoughts? Probably needs a fix?

**Comment by tacomanator** [comment]
I believe Haskell's implementation is correct.

Don't know the background on why parameter order was reversed in #1975, but the tuple order was  not changed at the same time. If this was intentional, then the docs are incorrect. Otherwise the tuple order needs updating:

```javascript
var mapAccumRight = _curry3(function mapAccumRight(fn, acc, list) {
  var idx = list.length - 1;
  var result = [];
  var tuple = [acc];                  // [[], acc]
  while (idx >= 0) {
    tuple = fn(list[idx], tuple[0]);  // fn(list[idx], tuple[1])
    result[idx] = tuple[1];           // tuple[0]
    idx -= 1;
  }
  return [result, tuple[0]];          // [result, tuple[1]]
});
```

Seems to make sense that the parameter and tuple order would match.

**Comment by kedashoe** [comment]
Would you like to submit a PR? I don't think we should deviate from what Haskell does here.

**Comment by CrossEye** [comment]
@paledpind: Are you ok with this change?  You were the one who suggested this for consistency reasons in https://github.com/ramda/ramda/pull/1975#issuecomment-261180741.'

And how about you @adispring?

**Comment by tacomanator** [comment]
Sure.

To be clear I would leave the argument order consistent with `reduceRight`, which is opposite Haskell's order, but correct the result. It would be a breaking change for any code relying on the current (incorrect) ordering.

```javascript
R.mapAccumRight((val, acc) => [val + acc, acc + "x"], "x", ["a", "b", "c"])
//=> [["axxx", "bxx", "cx"], "xxxx"]
```

Would be happy to submit PR once there is consensus.

**Comment by adispring** [comment]
Is `reduceRight` arguments' order incorrect?, it's consistent with Haskell's `foldr`.

**Comment by tacomanator** [comment]
I believe `reduceRight` arguments' order is correct and consistent with Haskell, but Haskell uses the opposite order for their `mapAccumR`. So, should `mapAccumRight` be consistent with `reduceRight` or Haskell's `mapAccumR`? @adispring

**Comment by adispring** [comment]
Yes, my refactored `R.mapAccumRight` implementation is wrong (https://github.com/ramda/ramda/pull/1975/files), I only reversed the `fn`'s arguments order, but forgot to reverse the tuple items` order

The implementation of `mapAccumRight you fixed is correct:

```js
var mapAccumRight = _curry3(function mapAccumRight(fn, acc, list) {
  var idx = list.length - 1;
  var result = [];
  var tuple = [acc];                  // [[], acc]
  while (idx >= 0) {
    tuple = fn(list[idx], tuple[0]);  // fn(list[idx], tuple[1])
    result[idx] = tuple[1];           // tuple[0]
    idx -= 1;
  }
  return [result, tuple[0]];          // [result, tuple[1]]
});
```

I think `R.mapAccumRight` should be consistent with `R.reduceRight`?  Haskell's `mapAccumR` arguments order is not consistent with `foldr`.

**Comment by tacomanator** [comment]
Did a little digging.

I asked on StackOverflow, but [according to the comments](https://stackoverflow.com/questions/47370404/what-the-haskell-language-design-teams-opinion-as-to-why-foldr-and-mapaccumr-us#comment81695029_47370404) it is [unlikely to be answered](https://stackoverflow.com/questions/44558242/picture-how-mapaccumr-works/44564452#44564452).

So I looked through the Haskell specs and it seems that `mapAccumL` and `mapAccumR` were added at the same time in [Haskell 1.4](http://haskell.org/definition/haskell-library-1.4.ps.gz) (around 1998). My guess is that since `mapAccumR` was described as "similar to `mapAccumL` except that the list is processed from right to left rather than left to right", they were implemented with the same signature and `foldr` was not taken into account or judged to be less similar.

It's a judgement call. Fixing the implementation of #1975 may break existing code, and there doesn't seem to be a strong opinion one way or the other. Thus, as it's probably less disruptive, what if I were to revert the change to `mapAccumRight` made in #1975 and fix the documentation?
